### DISCUSSION TITLE
Software in Support of CCO and CCO-based ontologies (a wish list)

### DISCUSSION THREAD

**Discussion by jonathanvajda** [CommonCoreOntology/CommonCoreOntologies#347] [General]
### CCO Developer Shopping List
I thought it would be worth having a list of software that does NOT exist yet to share amongst the community, but is earnestly desired. Software that could run by itself to help create, update, transform, display, or otherwise manage ontologies. Or the software might be an add on, a way to improve software we use such as Protégé plugins or MS Visual Studio Code extensions.

This is not a request from CCO developers to make these or support them. But we have common interests and similar workflows, and so we often have a common set of tools.

This is a crowd-sourced list, and if you add to the discussion make sure you give some functional requirements, not just a title.

When the list is fulfilled, I'll cross it out and link to that GitHub repo or website where it is accessible.

## Wishlist
- [ ] **Ontology spell checker**. Checks the literal values of all annotation type properties (rdfs:label, skos:prefLabel, skos:altLabel, skos:definition, etc.) for typos and/or grammatical errors. Generates a report what needs to be changed for each IRI. Stretch goal: gets user feedback in real-time to approve or reject suggested changes, creates a fixed ontology file with OWL API, and a report on what changes were made.
- [ ] **SPARQL to visual diagram generator**. Takes SPARQL query as input, and as output generates a visual diagram of the edges and nodes. Displays rdfs:label or skos:prefLabel wherever available. Stretch goal: exports as mermaid diagram file.
- [ ] **SPARQL _result_ to visual diagram generator**. Takes SPARQL query _results_ as input, and as output generates a visual diagram of edges and nodes. Displays rdfs:label or skos:prefLabel wherever available. Stretch goal: exports as mermaid diagram file.
- [ ] **SHACL to visual diagram generator.** Takes SHACL constraint as input, and as output generates a visual diagram of the edges, nodes, restrictions, etc. Displays rdfs:label or skos:prefLabel wherever available. Stretch goal: exports as mermaid diagram file.
- [ ] **Visual diagram to SPARQL generator**. Takes visual diagram of edges, nodes, etc. as input, and as output generates a SPARQL query.
- [ ] **Visual diagram to SHACL generator**. Takes visual diagram of edges, nodes, etc. as input, and as output generates SHACL constraints.
- [ ] **GUI Wrapper of RDFLib**
- [ ] **GUI wrapper for owlready2**
- [X] **GUI wrapper for ROBOT** [https://github.com/UTHealth-Ontology/GOBOT/](https://github.com/UTHealth-Ontology/GOBOT/)
- [ ] **OWL2 to Prolog converter**. Takes TTL, RDF-XML, or other OWL file as input, and as output generates a prolog file.
- [ ] **JSON-LD validator**. Not all JSON-LD files can be serialized into a valid OWL2 file. Checks whether the file can be converted without changing semantics or having gaps (e.g., every predicate of a triple has an IRI).

That said, I think it is worth having a markdown of actual software that supports ontology development, in support of this discussion https://github.com/CommonCoreOntology/CommonCoreOntologies/discussions/322 .

This is just the wishlist. Other ideas of 'gosh I need this all the time' for CCO-based ontology development? Please add!

**Comment by swartik** [comment]
@jonathanvajda TopBraid Composer does some of what you wish for. It can [generate SPARQL from visual diagrams](https://topbraidcomposer.org/html/). I think it can generate visual diagrams from SPARQL and SHACL results too. I haven't used it for ages.
 
The problem with TopBraid Composer is that it's not cheap. So, more general question. Do you want to include commercial products on this list?

**Reply by jonathanvajda** [reply]
I think I am looking for freeware or open source.

I haven't used TBC for a while. Is this available in the free version (not public anymore, as far as I know)?

**Comment by alanruttenberg** [comment]
Another: To support easier editing of SHACL constraints, the ability to mark, in Protege, that an axiom should be considered a SHACL constraint rather than an OWL one, and then a tool used during build that strips those out and generates the appropriate SHACL. Not all SHACL can be expressed in OWL, but many OWL axioms can be expressed (albeit with slightly different semantics) in SHACL.
